exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,419 @@
+ */
+
+use exonum::{
+    api::ApiContext,
+    blockchain::Schema as CoreSchema,
+    crypto::{Hash, PublicKey, SecretKey},
+    helpers::ValidatorId,
+    messages::BinaryValue,
+    exonum_merkledb::{self, Fork, Snapshot},
+    node::ApiSender,
+    runtime::{
+        api::ServiceApiBuilder,
+        dispatcher::{Dispatcher, DispatcherRef, DispatcherSender},
+        ArtifactId, ArtifactProtobufSpec, CallInfo, ErrorKind, ExecutionContext, ExecutionError,
+        InstanceDescriptor, InstanceId, InstanceSpec, Runtime, RuntimeIdentifier,
+        StateHashAggregator,
+    },
+};
+use futures::{Future, IntoFuture};
+use jni::{
+    objects::{GlobalRef, JObject, JValue},
+    signature::{JavaType, Primitive},
+    Executor,
+};
+use proto;
+use proxy::node::NodeContext;
+use runtime::Error;
+use std::fmt;
+use storage::View;
+use to_handle;
+use utils::{jni_cache::runtime_adapter, panic_on_exception, unwrap_jni};
+use JniErrorKind;
+use JniResult;
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+}
+
+/// Artifact identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct JavaArtifactId(String);
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceStateHashes")]
+struct ServiceStateHashes {
+    instance_id: u32,
+    state_hashes: Vec<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceRuntimeStateHashes")]
+struct ServiceRuntimeStateHashes {
+    runtime_state_hashes: Vec<Vec<u8>>,
+    service_state_hashes: Vec<ServiceStateHashes>,
+}
+
+impl JavaRuntimeProxy {
+    /// Runtime Identifier
+    pub const RUNTIME_ID: RuntimeIdentifier = RuntimeIdentifier::Java;
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            Err(Error::IncorrectArtifactId.into())
+        } else {
+            Ok(JavaArtifactId(artifact.name.to_string()))
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
+        res.map_err(|err| {
+            let kind: ErrorKind = match err.kind() {
+                JniErrorKind::JavaException => Error::JavaException.into(),
+                _ => Error::UnspecifiedError.into(),
+            };
+            (kind, err).into()
+        })
+    }
+
+    /// If the current node is a validator, returns `Some(validator_id)`, for other nodes return `None`.
+    fn validator_id(snapshot: &dyn Snapshot, pub_key: &PublicKey) -> Option<ValidatorId> {
```

Comment:
The code obtaining the ValidatorId is (still) unnecessarily complex. 

@alekseysidorov Why has the validator id been removed from afterCommit arguments? Shan't the core then provide an easy way to obtain it? Or shall we remove it as well?

Reply:
@alekseysidorov 🏓

Reply:
In old implementation after commit context contains parsed blockchain configuration so seems a little bit redundant in new code. Without this configuration impossible to get validator ID, so you have to get it explicitily from information schema to get validator ID.

Reply:
Shall the runtime provide the validator id in after_commit? Do services need to distinguish if they are validators or auditors, which the validator id enables to do?

Reply:
Runtime should not provide any consensus configuration and similar things so as not to break encapsulation.
In the end, the consensus algorithm may be different.

Reply:
@alekseysidorov @popzxc 

Does Exonum keep supporting submitting txs in services? If I understand correctly, services running on auditors must not generally do that. How are they supposed to determine if they run on auditors or not?